README.md
# MultiversX Terraform Stack for Observer node types

This solution deploys a Terraform stack for the MultiversX observer nodes on Tencent Cloud. It uses Tencent Cloud Lighthouse as main compute service alongside other necessary resources.
The chosen Operating system is an OpenCloudOS (centOS compatible) with docker pre-installed.

Multiversx documentation: https://github.com/multiversx/mx-chain-observing-squad#observing-squad

# Choose the right observer node type 
Here are the options.

**Observer-Lite**: MultiversX Observer node with "lite" deployment mode for all 4 shards, which optimizes speed but doesn't offer access to the full blockchain history but only stores the current epoch. Usage:
 - light dApps which don't need full access to all blockchain history
 - suited for high load API calls for basic ops such as getAddress, getBalance, listAssets, etc 
 - optimized for account queries and smart contract queries

**Observer-DBlookup-HDD**: MultiversX Observer node with "db-lookup" deployment mode for all 4 shards, which balances sync speed and access to full chains history. Usage:
 - most of the dApps which need to balance speed and history
 - includes the dblookup option so can have extended block db history
 - Separate Premium HDD drives for every shard delivering up to 6000iops per disk

**Observer-DBlookup-SSD**: MultiversX Observer node with "db-lookup" deployment mode for all 4 shards, which balances sync speed and access to full chains history. Usage:
 - most of the dApps which need to balance speed and history
 - includes the dblookup option so can have extended block db history
 - Separate SSD drives for every shard delivering up to 12000iops per disk


# Features

 - The underlying compute service used is called [Tencent Cloud Lighthouse](https://www.tencentcloud.com/products/lighthouse)
 - The Cloud Accounts creation process will be streamlined for speed
 - There is an Account preparation phase which needs to be made prior to node deployment
 - The node deployment repository will be made available alongside the procedure for deployment
 - The accounts and instances are secured according Cloud and industry best practices
 - For the DB-LOOKUP nodes, their up-to-date (max 24h) snapshot is retrieved before deployment
 - The deployment of the shard processes is made in docker containers in separate “screens” (to make debugging and operation easy)
 - Integrated instance resource monitor is available in Lighthouse panel
 - All Instances have Tencent Cloud CLI and COS tools pre-installed
 - The code also includes update / delete commands

# Pre-requisites (procedures)

## Tencent Cloud Account Creation and Setup

Please follow the below procedures to set-up your Tencent Cloud Account
 - [Sign up for Tencent Cloud Account](https://www.tencentcloud.com/document/product/378/17985)
 - [Add and verify the mobile number](https://www.tencentcloud.com/document/product/378/48918)
 - [Log in with your newly created root account](https://www.tencentcloud.com/document/product/378/36004)
 - [Verify the account owner](https://www.tencentcloud.com/document/product/378/3629) (individual or enterprise)
 - [Enable login protection](https://www.tencentcloud.com/document/product/378/8392) and [operation protection](https://www.tencentcloud.com/document/product/378/10740)
 - [Enable suspicious login protection](https://www.tencentcloud.com/document/product/378/10740)
 - [Bind MFA device for login and operation](https://www.tencentcloud.com/document/product/378/32528)
 - [Manage a secure password](https://www.tencentcloud.com/document/product/378/14623)
 - [Create an administrator sub-user](https://www.tencentcloud.com/document/product/598/38247)
 - For the newly created sub-user, go to CAM -> User List -> “username” and [enable MFA](https://www.tencentcloud.com/document/product/378/32528). This is different account / device from the root user MFA
 - Make sure to follow [other best practices](https://www.tencentcloud.com/document/product/598/10592) for enhancing the security

## Deployment configuration
The deployment is made with terraform, directly through the API of the Tencent Cloud Account created in the step above. To achieve the deployment, the environment must be set-up. Here are the steps:

### Step1 - Generate new Tencent Cloud API keys and export them locally

**MacOS & Linux**
 - Create a new folder containing your project and switch to it 
      `mkdir $HOME/mvx-observer && cd "$_"`
 - Create a new environment variable file that will be used for sourcing (e.g `nano vars.txt`). This file should contain:
      `"TENCENTCLOUD_SECRET_ID": "",`
      `"TENCENTCLOUD_SECRET_KEY": "",`
      `"TENCENTCLOUD_REGION": ""`
   - The first 2 values you need to paste from the Tencent cloud console Access Key section and the region must be taken from the [valid list of regions](https://www.tencentcloud.com/document/product/416/6479?lang=en), for example `eu-frankfurt`
   - Once the 3 lines are added to the text file, save and close (**CTRL+X & Y** for nano)
   - Remember that this file contains the credentials to your Tencent Cloud Account, so it must be secured and not to be disclosed to unauthorized parties
 - source the file above to write the variables to current environment
      `source ~/Documents/Projects/export.txt` (example)
Once the environment variables are set you can continue to the next step

### Step2 - configure local deployment environment

 - [Install VSCode](https://code.visualstudio.com/)
 - Install VSCode Terraform extension
 - Open the terminal and Install terraform locally. For example on MacOS
     - `brew tap hashicorp/tap`
     - `brew install hashicorp/tap/terraform`
     - `terraform -version`

## Solution deployment

If you got up to this point, you are ready to deploy the solution.
For this you need to create a single file, which will contain the module instantiation with the custom parameters

Here is a sample

### Sample lite observer
```hcl
module "multiversx-observer-lite" {
  
#-------source repo
  source = "tencentcloud-tci/multiversx-node/tencentcloud" #terraform module published in the registry
  
  #-------basic variables
  az            = "eu-frankfurt-1" #availability zone to deploy
  instance_name = "mvx-myobserver" #name of the LH instance
  
  #-------deployment variables
  deployment_mode = "lite" #the deployment mode: lite, db-lookup-hdd, db-lookup-ssd
  purchase_period = 1 #the valability of the purchase, in months
  blueprint_id  = "lhbp-a7oxy3em" #the id of the blueprint, we recommend using OpenCloudOS8 with docker

  
  #-------State specific variables
  need_tat_commands = true #set 'false' only if the commands are already deployed (if previous/paralel deployment existed)
  
  #-------Chain specific variables
  network = "mainnet" #choose between mainnet, testnet, devnet
  
  #-------firewall details
  ssh_client_cidr = "2.222.22.2/32" #source ip of the management location (for SSH whitelisting)
  extra_firewall_rules = [{ #specify the public proxy port
    
      protocol                  = "TCP"
      port                      = "8079"
      cidr_block                = "0.0.0.0/0"
      action                    = "ACCEPT"
      firewall_rule_description = "proxy port"
    
  }]
}
```
### Sample db-lookup observer
```hcl
module "multiversx-observer" {
  
  #-------source repo
  source = "tencentcloud-tci/multiversx-node/tencentcloud" #terraform module published in the registry
  
  #-------basic variables
  az            = "eu-frankfurt-1" #availability zone to deploy
  instance_name = "mx-myobserver" #name of the LH instance
  
  #-------deployment variables
  deployment_mode = "lite" #the deployment mode: lite, db-lookup-hdd, db-lookup-ssd
  purchase_period = 1 #the valability of the purchase, in months
  blueprint_id  = "lhbp-a7oxy3em" #the id of the blueprint, we recommend using OpenCloudOS8 with docker
  
  #-------State specific variables
  need_tat_commands = true #set 'false' only if the commands are already deployed (if previous/paralel deployment existed)
  
  #-------Chain specific variables
  network = "mainnet" #choose between mainnet, testnet, devnet

  #-------firewall details
  ssh_client_cidr = "2.222.22.2/32" #source ip of the management location (for SSH whitelisting)
  extra_firewall_rules = [{ #specify the public proxy port
    
      protocol                  = "TCP"
      port                      = "8079"
      cidr_block                = "0.0.0.0/0"
      action                    = "ACCEPT"
      firewall_rule_description = "proxy port"
    
  }]
  #-------disk variables for db-lookup option
  #leave default unless disk becomes full
  cbs0_disk_size = 350 #disk contains node-0 and node-metachain data
  cbs1_disk_size = 450 #disk contains node-1
  cbs2_disk_size = 300 #disk contains node-2
  floating_cbs = "lhdisk-jkx4d2w4" #ID of the floater disk which will be used to download and extract the node DB history
}
```

### Main parameters
A few of parameters are needed to deploy the instance
 - `source` is the terraform registry module name
 - `az` is the availability zone within the selected region
 - `instance_name` give a name to your node
 - `purchase_period` the purchase period in months
 - `deployment_mode` the deployment mode: lite, db-lookup-hdd, db-lookup-ssd
 - `bundle_id` used to select the instance type, leave default. [Other bundles](https://www.tencentcloud.com/document/product/1103/42472)
 - `blueprint_id` used for selection of the instance image, leave default. [Other blueprints](https://www.tencentcloud.com/document/product/1103/42503)
 - `ssh_client_cidr` source ip of the management location (for SSH whitelisting)
 - `need_tat_commands` set 'false' only if the commands are already deployed (if previous/paralel deployment existed)
 - `network` choose between mainnet, testnet, devnet. This selection will switch the MultiversX network.
 - `floating_cbs` ID of the floater disk which will be used to download and extract the node DB history
 - `extra_firewall_rules` define own custom firewall inbound rule
 - `cbsX_disk_size` where X=0,1,2. this variable defines the size of the disk for the node databases
For reference here is the [Tencent Cloud Terraform provider](https://registry.terraform.io/providers/tencentcloudstack/tencentcloud/latest/docs/resources/lighthouse_instance)

### Stack deployment
Having the configuration done, continue with these commands
 - `terraform init`
 - `terraform plan`
 - `terraform apply` select yes, enter

The deployment will take more the for db-archive node types because of the node database which has to be retrieved from the mirror

If you are deploying in the same folder an updated version of the stack please upgrade the modules from source with the command below:
- `terraform init -upgrade`
More details [here](https://developer.hashicorp.com/terraform/cli/commands/init#upgrade)


## Deployment details

We use screen when start the different services. There are 5 screen sessions created:

* proxy
* squad-metachain
* squad-0
* squad-1
* squad-2

If you login the console hit `screen -r` to view the running screens and if you wish to enter one of the shard or proxy processes add its name at the end. For example `screen -r squad-1`
To exit the screen mode hit `CTRL-A + D`

Here are some details of the supported modes:
### Lite node type

| Program | Directory |
| -- | -- |
| node-0 | /data/MyObservingSquad/node-0 |
| node-1 | /data/MyObservingSquad/node-1 |
| node-2 | /data/MyObservingSquad/node-2 |
| node-metachain | /data/MyObservingSquad/node-metachain |
| proxy | /data/MyObservingSquad/proxy |


### DB-lookup node type
This node type requires an additional 3 cloud disks: cbs-0, cbs-1, cbs-2 plus one temporary disk 'cbs_float'

* cbs-0 containes the deployment of: node-0 and node-metachain
* cbs-1 containes the deployment of: node-1
* cbs-2 containes the deployment of: node-2
* cbs_float is a temporary disk, used to download the block database archives. By using these archive files, we can speed up the progress of synchronization during the initial deployment. When the node deployment is done, this temporary disk will be detached from the node.

| Service | Directory |
| -- | -- |
| node-0 | /data/MyObservingSquad/cbs-0/node-0 |
| node-1 | /data/MyObservingSquad/cbs-1/node-1 |
| node-2 | /data/MyObservingSquad/cbs-2/node-2 |
| node-metachain | /data/MyObservingSquad/cbs-0/node-metachain |
| proxy | /data/MyObservingSquad/proxy |

For the db-lookup option there are types of cloud disk available:

* hdd: corresponds to "premium cloud disks" on TencentCloud
* ssd: corresponds to "SSD cloud disks" on TencentCloud

Check the performance of these 2 cloud disk types [here](https://www.tencentcloud.com/document/product/362/31636)

## How to use the TAT commands
### multiversx-node-runner
This command is used to deploy a node. 

#### Parameters
`deployment_mode`
* required
* value: lite, db-lookup-hdd, db-lookup-ssd

when `deployment_mode=db-lookup-hdd, db-lookup-ssd`, extra variables should be set:

* `floating_cbs`: the temporary disk ID
* `deployment_mode`: has to be one of the 2 db-lookup options


### multiversx-node-tool
This command is for routine operations.

#### Parameters
`command`
* required
* value: upgrade_all, stop_all, start_all

| Value | Desc |
| -- | -- |
| `upgrade_all` | To check the docker image(multiversx/chain-observer, multiversx/chain-squad-proxy) latest tag. If there is a new tag, it will download and restart the node. |
| `stop_all` | To stop the node's programs |
| `start_all` | To start the node's programs |

# Managing the node

## Logging in
 - Log in Tencent Cloud Console and go to Lighthouse service 
 - Identify the instance and click log in -> Auto -> Login. you are now in the console

## Verify operation
 - Verify the node processes are running. You should see something like below
 ```bash
          [lighthouse@VM-0-24-centos ~]$ sudo -i
          [root@VM-0-24-centos ~]# docker ps
          CONTAINER ID   IMAGE                                COMMAND                  CREATED              STATUS              PORTS                                NAMES
          d99f497501cd   multiversx/chain-squad-proxy:using   "./proxy ./proxy"        About a minute ago   Up About a minute   8079-8080/tcp                        proxy
          1ade927d96e7   multiversx/chain-observer:using      "/go/mx-chain-go/cmd…"   About a minute ago   Up About a minute   8080/tcp, 0.0.0.0:10003->37373/tcp   squad-metachain
          c20f7bc91887   multiversx/chain-observer:using      "/go/mx-chain-go/cmd…"   About a minute ago   Up About a minute   8080/tcp, 0.0.0.0:10002->37373/tcp   squad-2
          51c1aa339df2   multiversx/chain-observer:using      "/go/mx-chain-go/cmd…"   About a minute ago   Up About a minute   8080/tcp, 0.0.0.0:10001->37373/tcp   squad-1
          30a5d0c28e48   multiversx/chain-observer:using      "/go/mx-chain-go/cmd…"   About a minute ago   Up About a minute   8080/tcp, 0.0.0.0:10000->37373/tcp   squad-0
          [root@VM-0-24-centos ~]# screen -r
          There are several suitable screens on:
               4481.proxy      (Detached)
               4420.squad-metachain    (Detached)
               4378.squad-2    (Detached)
               4357.squad-1    (Detached)
               4332.squad-0    (Detached)
          Type "screen [-d] -r [pid.]tty.host" to resume one of them.
          [root@VM-0-24-centos ~]# 
```
 - There are other, more advanced ways to verify node operation. Please check [MultiversX repository](https://github.com/multiversx/mx-chain-observing-squad)
 - Verify machine load and traffic by clicking the `Monitoring` tab
 



CHANGELOG.md
# Changelog

## [0.4.0](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.4.0) (2023-11-18)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.3.3...0.4.0)

**Merged pull requests:**

- feat: support devnet and testnet [\#2](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/pull/2) ([ritch2022](https://github.com/ritch2022))

## [0.3.3](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.3.3) (2023-10-17)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.3.2...0.3.3)

**Merged pull requests:**

- fix readme and example [\#1](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/pull/1) ([tudorpaultoma](https://github.com/tudorpaultoma))

## [0.3.2](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.3.2) (2023-10-10)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.3.1...0.3.2)

## [0.3.1](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.3.1) (2023-08-22)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.3.0...0.3.1)

## [0.3.0](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.3.0) (2023-08-16)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.2.6...0.3.0)

## [0.2.6](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.2.6) (2023-07-24)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.2.5...0.2.6)

## [0.2.5](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.2.5) (2023-07-06)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.2.4...0.2.5)

## [0.2.4](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.2.4) (2023-06-25)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.2.3...0.2.4)

## [0.2.3](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.2.3) (2023-06-22)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.2.2...0.2.3)

## [0.2.2](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.2.2) (2023-06-21)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.2.1...0.2.2)

## [0.2.1](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.2.1) (2023-06-13)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.2.0...0.2.1)

## [0.2.0](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.2.0) (2023-06-08)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.1.3...0.2.0)

## [0.1.3](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.1.3) (2023-04-11)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.1.2...0.1.3)

## [0.1.2](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.1.2) (2023-03-21)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/0.1.1...0.1.2)

## [0.1.1](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/tree/0.1.1) (2023-03-21)

[Full Changelog](https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node/compare/9cd57218c0a78b23cc501a9957840c87b42019b7...0.1.1)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


scripts/squad-run-node.sh
#!/bin/bash

# this script is suitable for centOS based distributions

set -e
SQUAD_BASE_DIR=/data/MyObservingSquad
FLOAT_MOUNT_DIR=/data/float
PROXY_BASE_DIR=$SQUAD_BASE_DIR/proxy
DEPLOYMENT_MODE_FILE=$SQUAD_BASE_DIR/.deployment_mode
DOCKER_IMAGE_NODE_NAME_FILE=$SQUAD_BASE_DIR/.node_image_name

KEY_DIR=/data/keys

REGION=`curl http://metadata.tencentyun.com/latest/meta-data/placement/region`
LH_ID={{lighthouse_id}}
CBS_ID_0={{cbs_0}}
CBS_ID_1={{cbs_1}}
CBS_ID_2={{cbs_2}}
CBS_ID_FLOAT={{cbs_float}}
NETWORK={{network}}

#
export TENCENTCLOUD_SECRET_ID={{secret_id}}
export TENCENTCLOUD_SECRET_KEY={{secret_key}}
export TENCENTCLOUD_REGION=$REGION

set_docker_image_name() {
    case $NETWORK in
        "mainnet")
            DOCKER_IMAGE_NODE_NAME="chain-observer"
            ;;
        "testnet")
            DOCKER_IMAGE_NODE_NAME="chain-testnet"
            ;;
        "devnet")
            DOCKER_IMAGE_NODE_NAME="chain-devnet"
            ;;
    esac
}

pull_docker_images() {
    set_docker_image_name

    local dhgenkey="https://registry.hub.docker.com/v2/repositories/multiversx/chain-keygenerator/tags"
    local dhobs="https://registry.hub.docker.com/v2/repositories/multiversx/${DOCKER_IMAGE_NODE_NAME}/tags"
    local dhpro="https://registry.hub.docker.com/v2/repositories/multiversx/chain-squad-proxy/tags"
    
    yum install -y -q jq
    local latgenkey=`curl -s -S $dhgenkey | jq '."results"[]["name"]' | sed -n '1p' | tr -d '"'`
    local latobs=`curl -s -S $dhobs | jq '."results"[]["name"]' | sed -n '1p' | tr -d '"'`
    local latpro=`curl -s -S $dhpro | jq '."results"[]["name"]' | sed -n '1p' | tr -d '"'`

    docker pull -q multiversx/chain-keygenerator:$latgenkey
    docker tag multiversx/chain-keygenerator:$latgenkey multiversx/chain-keygenerator:using
    docker pull -q multiversx/${DOCKER_IMAGE_NODE_NAME}:$latobs
    docker tag multiversx/${DOCKER_IMAGE_NODE_NAME}:$latobs multiversx/${DOCKER_IMAGE_NODE_NAME}:using
    docker pull -q multiversx/chain-squad-proxy:$latpro
    docker tag multiversx/chain-squad-proxy:$latpro multiversx/chain-squad-proxy:using
}

# input:
#   $1: dir to store the key file
#   $2: name of the key file
generate_key() {
    if [ -f "$1/$2" ]; then
        echo "===== Skipping key generation ====="
    else
        echo "===== Generating key ====="
        docker run --privileged --rm --mount type=bind,source=$1,destination=/keys --workdir /keys multiversx/chain-keygenerator:using \
        && sudo chown $(whoami) $1/validatorKey.pem \
        && mv $1/validatorKey.pem $1/$2
    fi
}

run() {
    local SHARD=$1
    local DISPLAY_NAME=$2
    local OBSERVER_DIR=$3
    local P2P_PORT=$4
    local IP=$5
    if [ -z $(docker ps -q -f "name=squad-$SHARD") ]; then
        echo "===== Running node $SHARD ====="
        if [ "$SHARD" == "metachain" ]; then
            screen -dmS squad-${SHARD} docker run --privileged --rm \
            --mount type=bind,source=${OBSERVER_DIR}/db,destination=/go/mx-chain-go/cmd/node/db \
            --mount type=bind,source=${OBSERVER_DIR}/logs,destination=/go/mx-chain-go/cmd/node/logs \
            --mount type=bind,source=${OBSERVER_DIR}/config,destination=/config \
            --publish ${P2P_PORT}:37373 --network=multiversx-squad --ip=${IP} \
            --name squad-${SHARD} multiversx/${DOCKER_IMAGE_NODE_NAME}:using \
            --destination-shard-as-observer=${SHARD} \
            --validator-key-pem-file=/config/observerKey_${SHARD}.pem --display-name="${DISPLAY_NAME}" \
            --operation-mode db-lookup-extension

        elif [ "$6" == "lite" ]; then
            screen -dmS squad-${SHARD} docker run --privileged --rm \
            --mount type=bind,source=${OBSERVER_DIR}/db,destination=/go/mx-chain-go/cmd/node/db \
            --mount type=bind,source=${OBSERVER_DIR}/logs,destination=/go/mx-chain-go/cmd/node/logs \
            --mount type=bind,source=${OBSERVER_DIR}/config,destination=/config \
            --publish ${P2P_PORT}:37373 --network=multiversx-squad --ip=${IP} \
            --name squad-${SHARD} multiversx/${DOCKER_IMAGE_NODE_NAME}:using \
            --destination-shard-as-observer=${SHARD} \
            --validator-key-pem-file=/config/observerKey_${SHARD}.pem --display-name="${DISPLAY_NAME}" \
            --operation-mode=snapshotless-observer
        else
            screen -dmS squad-${SHARD} docker run --privileged --rm \
            --mount type=bind,source=${OBSERVER_DIR}/db,destination=/go/mx-chain-go/cmd/node/db \
            --mount type=bind,source=${OBSERVER_DIR}/logs,destination=/go/mx-chain-go/cmd/node/logs \
            --mount type=bind,source=${OBSERVER_DIR}/config,destination=/config \
            --publish ${P2P_PORT}:37373 --network=multiversx-squad --ip=${IP} \
            --name squad-${SHARD} multiversx/${DOCKER_IMAGE_NODE_NAME}:using \
            --destination-shard-as-observer=${SHARD} \
            --validator-key-pem-file=/config/observerKey_${SHARD}.pem --display-name="${DISPLAY_NAME}" \
            --operation-mode db-lookup-extension
        fi
    else
        echo "===== INFO: the node $SHARD is already running ====="
    fi
}

init_env() {
    echo "===== Initialising environment ====="
    echo "===== Installing epel-release ... ====="
    yum -y -q install epel-release
    echo "===== Installing python / screen / tccli ... ====="
    yum install -y -q python3 screen
    python3 -m pip install pip
    pip3 install -q tccli
    pip3 install --upgrade tccli
    cp /usr/local/bin/tccli /usr/bin/

}

init_dir_lite() {
    NODE_0_DIR=$SQUAD_BASE_DIR/node-0
    NODE_1_DIR=$SQUAD_BASE_DIR/node-1
    NODE_2_DIR=$SQUAD_BASE_DIR/node-2
    META_NODE_DIR=$SQUAD_BASE_DIR/node-metachain

    mkdir -p $SQUAD_BASE_DIR/{proxy,node-0,node-1,node-2,node-metachain}/{config,logs}
    mkdir -p $SQUAD_BASE_DIR/{node-0,node-1,node-2,node-metachain}/db

    generate_key "$NODE_0_DIR/config" "observerKey_0.pem"
    generate_key "$NODE_1_DIR/config" "observerKey_1.pem"
    generate_key "$NODE_2_DIR/config" "observerKey_2.pem"
    generate_key "$META_NODE_DIR/config" "observerKey_metachain.pem"
}

init_env_db-lookup() {
    echo "===== Initialising environment for db-lookup extension ====="
    echo "LH_ID=$LH_ID"
    echo "CBS_ID_0=$CBS_ID_0"
    echo "CBS_ID_1=$CBS_ID_1"
    echo "CBS_ID_2=$CBS_ID_2"
    echo "CBS_ID_FLOAT=$CBS_ID_FLOAT"
}

# input:
#   $1: cbs_id
#   $2: dir to mount
attach_and_mount() {
    if [[ -z `df | grep $2` ]]; then
        if [[ -n `ls -la /dev/disk/by-id/ |grep ${1#*-}` ]]; then
            echo "===== INFO: skipping attachament of $1 ====="
        else
            echo "===== Attaching $1 ====="
            tccli lighthouse AttachDisks --cli-unfold-argument --region $REGION --DiskIds $1 --InstanceId $LH_ID
            sleep 15
        fi
        local tmp=`ls -la /dev/disk/by-id/ |grep ${1#*-}`
        local cbs_dev=${tmp##*/}
        echo "===== Mounting cbs_dev=$cbs_dev to $2 ====="
        if [[ -n `parted -s /dev/$cbs_dev print 2>&1|grep -i error` ]]; then
            echo "===== Formatting $1 ====="
            mkfs -t ext4 /dev/$cbs_dev
        fi
        mkdir -p $2
        mount /dev/$cbs_dev $2

        if [ "$2" != "$FLOAT_MOUNT_DIR" ]; then
            echo "===== Configuring /etc/fstab for $1 ====="
            echo "/dev/disk/by-id/virtio-disk-${1#*-} $2 ext4 defaults 0 0" >> /etc/fstab
        fi
    else
        echo "===== INFO: $1 already mounted to $2 ====="
    fi
}


init_dir_db-lookup() {
    CBS_0_DIR=$SQUAD_BASE_DIR/cbs-0
    CBS_1_DIR=$SQUAD_BASE_DIR/cbs-1
    CBS_2_DIR=$SQUAD_BASE_DIR/cbs-2
    NODE_0_DIR=$CBS_0_DIR/node-0
    NODE_1_DIR=$CBS_1_DIR/node-1
    NODE_2_DIR=$CBS_2_DIR/node-2
    META_NODE_DIR=$CBS_0_DIR/node-metachain

    # create CBS

    # attach and mount the disks
    # format
    echo "===== Disk attaching and formatting ====="
    echo "TENCENTCLOUD_SECRET_ID=$TENCENTCLOUD_SECRET_ID"
    echo "TENCENTCLOUD_SECRET_KEY=$TENCENTCLOUD_SECRET_KEY"
    attach_and_mount $CBS_ID_0 $CBS_0_DIR
    attach_and_mount $CBS_ID_1 $CBS_1_DIR
    attach_and_mount $CBS_ID_2 $CBS_2_DIR
    attach_and_mount $CBS_ID_FLOAT $FLOAT_MOUNT_DIR


    # copy key files
    mkdir -p $NODE_0_DIR/{logs,config}
    generate_key "$NODE_0_DIR/config" "observerKey_0.pem"
    mkdir -p $NODE_1_DIR/{logs,config}
    generate_key "$NODE_1_DIR/config" "observerKey_1.pem"
    mkdir -p $NODE_2_DIR/{logs,config}
    generate_key "$NODE_2_DIR/config" "observerKey_2.pem"
    mkdir -p $META_NODE_DIR/{logs,config}
    generate_key "$META_NODE_DIR/config" "observerKey_metachain.pem"

    echo "===== Downloading data files ====="
    local ex="tgz"
    # download latest block DBs
    if [ ! -f $FLOAT_MOUNT_DIR/node-0.$ex ]; then
        wget -q https://tommyxyz-1301327510.cos.eu-frankfurt.myqcloud.com/MX/node-0.$ex -P $FLOAT_MOUNT_DIR
    fi
    if [ ! -f $FLOAT_MOUNT_DIR/node-1.$ex ]; then
        wget -q https://tommyxyz-1301327510.cos.eu-frankfurt.myqcloud.com/MX/node-1.$ex -P $FLOAT_MOUNT_DIR
    fi
    if [ ! -f $FLOAT_MOUNT_DIR/node-2.$ex ]; then
        wget -q https://tommyxyz-1301327510.cos.eu-frankfurt.myqcloud.com/MX/node-2.$ex -P $FLOAT_MOUNT_DIR
    fi
    if [ ! -f $FLOAT_MOUNT_DIR/node-metachain.$ex ]; then
        wget -q https://tommyxyz-1301327510.cos.eu-frankfurt.myqcloud.com/MX/node-metachain.$ex -P $FLOAT_MOUNT_DIR
    fi

    # extract block databases in parallel processes
    echo "===== Extracting database files ====="
    if [[ ! -d $NODE_0_DIR/db/db/1/Static || ! -d $NODE_0_DIR/db/1/Static ]]; then
        tar xf $FLOAT_MOUNT_DIR/node-0.$ex -C $NODE_0_DIR &
    fi
    if [[ ! -d $NODE_1_DIR/db/db/1/Static || ! -d $NODE_1_DIR/db/1/Static ]]; then
        tar xf $FLOAT_MOUNT_DIR/node-1.$ex -C $NODE_1_DIR &
    fi
    if [[ ! -d $NODE_2_DIR/db/db/1/Static || ! -d $NODE_2_DIR/db/1/Static ]]; then
        tar xf $FLOAT_MOUNT_DIR/node-2.$ex -C $NODE_2_DIR &
    fi
    if [[ ! -d $META_NODE_DIR/db/db/1/Static || ! -d $META_NODE_DIR/db/1/Static ]]; then
        tar xf $FLOAT_MOUNT_DIR/node-metachain.$ex -C $META_NODE_DIR &
    fi
    wait

}

cleanup() {
    # clean up
    echo "===== Cleaning up ====="
    if [ "{{deployment_mode}}" != "lite" ]; then
        umount $FLOAT_MOUNT_DIR
        tccli lighthouse DetachDisks --cli-unfold-argument --region $REGION --DiskIds $CBS_ID_FLOAT || true
    fi

    unset TENCENTCLOUD_SECRET_ID
    unset TENCENTCLOUD_SECRET_KEY
}

run_cis_hardening() {
    echo "===== Begin OS hardening ====="
    echo "===== 1. Installing ansible ... ====="
    sudo pip3 install ansible
    echo "===== 2. Installing git ... ====="
    yum -y -q install git
    echo "===== 3. Installing nss ... ====="
    yum -y -q install nss
    echo "===== 4. Cloning repo ... ====="
    git clone https://github.com/tencentcloud-tci/terraform-tencentcloud-multiversx-node.git /home/lighthouse/source-repo/
    cd /home/lighthouse/source-repo/
    echo "===== 5. Running ansible playbook for CIS OS hardening  ... ====="
    ansible-playbook /home/lighthouse/source-repo/scripts/cis-hardening/cis.yml > CIS-ansible.log
    echo "===== End OS hardening ====="
}

run_squad() {    
    if [ -f $DEPLOYMENT_MODE_FILE ]; then
        local curr_type=`cat $DEPLOYMENT_MODE_FILE`
        if [ "$1" != $curr_type ]; then
            echo "===== ERROR: deployment mode mismatch -> input=$1 current=$curr_type"
            exit 1
        else
            echo "===== INFO: squad already deployed, deployment mode: $1 ====="
            exit 0
        fi
    fi
    
    # init dir
    if [ "$1" == "lite" ]; then
        init_dir_lite
    elif [[ "$1" == "db-lookup-ssd" || "$1" == "db-lookup-hdd" ]]; then
        init_env_db-lookup
        init_dir_db-lookup
    else
        echo "===== ERROR: unsupported node type: $1 ====="
        exit 1
    fi

    # create a docker network
    if [ -z $(docker network ls -q -f "name=multiversx-squad") ]; then
        docker network create --subnet=10.0.0.0/24 multiversx-squad
    fi

    # Start Observer of Shard 0
    run 0 "MyObservingSquad-0" $NODE_0_DIR 10000 10.0.0.6 $1
    sleep 30

    # Start Observer of Shard 1
    run 1 "MyObservingSquad-1" $NODE_1_DIR 10001 10.0.0.5 $1
    sleep 30

    # Start Observer of Shard 2
    run 2 "MyObservingSquad-2" $NODE_2_DIR 10002 10.0.0.4 $1
    sleep 30

    # Start Observer of Metachain
    run metachain "MyObservingSquad-metachain" $META_NODE_DIR 10003 10.0.0.3 $1
    sleep 30

    # Start Proxy
    if [ -z $(docker ps -q -f "name=proxy") ]; then
        local IP=10.0.0.2
        screen -dmS proxy docker run --privileged --rm --network=multiversx-squad --ip=${IP} -p 8079:8079 --name proxy multiversx/chain-squad-proxy:using
    else
        echo "===== INFO: proxy service already running ====="
    fi

    # Write observer type
    echo "$1" > $DEPLOYMENT_MODE_FILE
    echo "$DOCKER_IMAGE_NODE_NAME" > $DOCKER_IMAGE_NODE_NAME_FILE

    cleanup
}

# ------------------------------
# main
# ------------------------------

echo "===== Deploying {{deployment_mode}} ====="

init_env

run_cis_hardening

pull_docker_images

run_squad {{deployment_mode}}
